Site Editor - extension point filter list for Site Editor init hook.

Addison-Stavlo opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Refactored the early return statement of the `gutenberg_site_editor_init` function to instead check a list that is filterable.  This will allow extra allowable hooks to be added by plugin developers.

Background:
We would like to enable the site editor experiment via a plugin and would like to add the 'Site Editor' button to the top-level of the sidebar.  Without updating the allowed hooks in the early return condition, this would be required to be a submenu page of the Gutenberg plugin page which we do not want shown in this context for various reasons.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Locally as well as with other plugin usage.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
added toplevel hook

epiqueras reviewed (commented):
I think it should be a configurable extension point, perhaps a filterable list.

Addison-Stavlo committed:
added a filter of filters

Addison-Stavlo committed:
updated filters to actually work o.O

Addison-Stavlo committed:
updated comments

Addison-Stavlo committed:
spell check... lol

Addison-Stavlo commented:
> I think it should be a configurable extension point, perhaps a filterable list.
Great!  That makes sense to me.  I have updated this to allow more customization:

* Added `add-site-editor-hook-path` filter to allow for setting custom hook strings allowed by the gutenberg_site_editor_init function.
* Also added the `is-ineligible-site-editor-hook` filter to be used within the gutenberg_site_editor_init function.
To set a new hook allowed by the init function one would now only need to: `apply_filters( 'add-site-editor-hook-path', $some_string)`

Addison-Stavlo committed:
newline at eof

Addison-Stavlo commented on lib/edit-site-page.php:
```diff
@@ -155,3 +155,24 @@ function gutenberg_edit_site_init( $hook ) {
+ * the init function itself.  This will filter through all hooks added by the outer filter and
+ * return 'false' if there is any match.  It will return the original hook if no match is found.
+ */
+add_filter( 'add-site-editor-hook-path', function( $new_hook ) {
```
Yeah that could work.  I guess the difference there is plugin developers will be expected to set up the correct callback function, where as here they don't need to set up anything other than the string to add.  Similarly with this we could:

```
add_action( 'add-site-editor-hook-path', function( $new_hook ) {
    add_filter( 'site-editor-hook-paths', function( $allowed_paths ) {
        return $allowed_paths[] = 'new_path';
    }
});

```
And then plugin developers could just `do_action('add-site-editor-hook-path', 'new_path')` not have to worry about defining the proper callback.

Addison-Stavlo committed:
refactored

vindl reviewed (approved):
LGTM! :shipit:

vindl commented on lib/edit-site-page.php:
```diff
@@ -34,7 +34,11 @@ function gutenberg_edit_site_init( $hook ) {
 		$_wp_current_template_part_ids;
-	if ( 'gutenberg_page_gutenberg-edit-site' !== $hook ) {
+
+	$allowed_hooks = [ 'gutenberg_page_gutenberg-edit-site', 'toplevel_page_gutenberg-edit-site' ];
```
Looking through the rest of the repo it seems like `array` is used instead of the shorthand syntax (likely for PHP backwards compatibility reasons). Since the second value of `toplevel_page_gutenberg-edit-site` is not really used in core now, what do you think about leaving it out?

vindl commented on lib/edit-site-page.php:
```diff
@@ -34,7 +34,11 @@ function gutenberg_edit_site_init( $hook ) {
-	if ( 'gutenberg_page_gutenberg-edit-site' !== $hook ) {
+
+	$allowed_hooks = [ 'gutenberg_page_gutenberg-edit-site', 'toplevel_page_gutenberg-edit-site' ];
+	$allowed_hooks = apply_filters( 'site-editor-hook-paths', $allowed_hooks );
```
Just noticed that the CI tests are failing because of this. The hook words need to be separated by underscores.

```suggestion
	$allowed_hooks = apply_filters( 'site_editor_allowed_hooks', $allowed_hooks );
```

Addison-Stavlo committed:
Update lib/edit-site-page.php

Co-Authored-By: Marko Andrijasevic <marko.andrijasevic@automattic.com>

Addison-Stavlo committed:
Update lib/edit-site-page.php

Co-Authored-By: Marko Andrijasevic <marko.andrijasevic@automattic.com>
